### PR TITLE
Add MinimalOboe Kotlin app

### PR DESCRIPTION
**Pull request by philburk**
Uses coroutine to call Oboe native to prevent ANR.

### CODE REVIEW

File: samples/minimaloboe/src/androidTest/java/com/example/minimaloboe/ExampleInstrumentedTest.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import org.junit.runner.RunWith
+
+import org.junit.Assert.*
+
+/**
+ * Instrumented test, which will execute on an Android device.
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+@RunWith(AndroidJUnit4::class)
```
Comment (robertwu1) [inline samples/minimaloboe/src/androidTest/java/com/example/minimaloboe/ExampleInstrumentedTest.kt:16]: Can we remove?

File: samples/minimaloboe/src/test/java/com/example/minimaloboe/ExampleUnitTest.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test
+
+import org.junit.Assert.*
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
```
Comment (robertwu1) [inline samples/minimaloboe/src/test/java/com/example/minimaloboe/ExampleUnitTest.kt:12]: Can we remove this test?

File: samples/minimaloboe/src/main/cpp/SimpleNoiseMaker.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+DataCallbackResult SimpleNoiseMaker::MyDataCallback::onAudioReady(
+        AudioStream *audioStream,
+        void *audioData,
+        int32_t numFrames) {
+    // We requested float when we built the stream.
+    float *output = (float *) audioData;
+    // Fill buffer with random numbers to create "white noise".
+    int numSamples = numFrames * kChannelCount;
+    for (int i = 0; i < numSamples; i++) {
+        // drand48() returns a random number between 0.0 and 1.0.
```
Comment (robertwu1) [inline samples/minimaloboe/src/main/cpp/SimpleNoiseMaker.cpp:74]: Can we play a sine wave instead so it is pleasant sounding?

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        return audioPlayer.stopAudio()
+    }
+}
+
+@Composable
+fun MainControls(viewModel: ExampleViewModel = ExampleViewModel()) {
+
+    // State that affects the UI.
+    var started by remember { mutableStateOf(false) }
+    var result by remember { mutableStateOf(1) }
```
Comment (robertwu1) [inline samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:69]: Having result start at 1 is quite strange. It doesn't map to Result::OK (0) or any of the error codes

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    Column {
+        Text(text = "Minimal Oboe!")
+        // START button
+        Button(
```
Comment (robertwu1) [inline samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:74]: nit: the button UI takes some time to run. Sound starts playing way before the UI is done updating. Do you know whether we can change the speed of the UI transitions so it takes less time?

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+class ExampleViewModel : ViewModel() {
+
+    var audioPlayer = AudioPlayer();
+
+    fun startAudio(): Int {
+        return audioPlayer.startAudio()
+
+    }
+
+    fun stopAudio(): Int {
```
Comment (robertwu1) [inline samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:59]: Should we stop audio if the screen rotates?

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/ui/theme/Type.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import androidx.compose.ui.unit.sp
+
+// Set of Material typography styles to start with
+val Typography = Typography(
+    body1 = TextStyle(
+        fontFamily = FontFamily.Default,
+        fontWeight = FontWeight.Normal,
+        fontSize = 16.sp
+    )
+    /* Other default text styles to override
```
Comment (robertwu1) [inline samples/minimaloboe/src/main/java/com/example/minimaloboe/ui/theme/Type.kt:16]: nit: remove comments?

File: samples/minimaloboe/src/main/cpp/MinimalOboeJNI.cpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+/**
+ * Native (JNI) implementation of AudioPlayer.stopAudioStreamNative()
+ */
+JNIEXPORT jint JNICALL Java_com_example_minimaloboe_AudioPlayer_stopAudioStreamNative(
+        JNIEnv * /* env */, jobject) {
+    __android_log_print(ANDROID_LOG_INFO, TAG, "%s", __func__);
+    Result result1 = sPlayer.stop();
+    Result result2 = sPlayer.close();
```
Comment (robertwu1) [inline samples/minimaloboe/src/main/cpp/MinimalOboeJNI.cpp:59]: Should we call close if stop fails?

File: samples/minimaloboe/src/main/cpp/SimpleNoiseMaker.cpp (hunk 3 above)
Comment (philburk) [reply samples/minimaloboe/src/main/cpp/SimpleNoiseMaker.cpp:74]: Some folks hate sine waves. And the sine wave requires state and a phaseIncrement. White Noise is more minimal.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 5 above)
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:74]: My buttons are fast on Pixel 7. Which device is slow?

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 6 above)
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:59]: I like that the sound continues. But the UI gets out of sync. Will fix.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 5 above)
Comment (robertwu1) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:74]: The UI takes about half a second for the full animation from purple -> gray -> black.

Comment (robertwu1) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:74]: This for Pixel 6.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 6 above)
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:59]: I also found the sound continues when the app is in the background. Not intended. So I added onPause(). It is better but when I switch tasks the UI can still get out of sync.

File: samples/minimaloboe/src/main/cpp/MinimalOboeJNI.cpp (hunk 8 above)
Comment (philburk) [reply samples/minimaloboe/src/main/cpp/MinimalOboeJNI.cpp:59]: Yes. The close() is needed to delete the resources.
If open() fails we do not start().
but if stop() fails we still want to close().
I added a comment.

**Comment by robertwu1** [comment]
Btw, I thought about Java samples a bit more and Soundboard and Megadrone may be too simple as examples. We simply start the audio engine when the app is resumed and stop the audio engine when the app is paused. This is good for many apps, but not all apps want to have an open audio stream all the time.

For Soundboard and Megadrone, I don't think we should change anything. Letting the UI thread open audio is fine in these cases because having pads that do not play audio at the beginning feels like a bug.

LiveEffect uses a different mechanism. Only when LiveEffectEngine.setEffectOn(bool) is called, should the engine be started or stopped. This seems like a good place to spin off another non-UI thread.

HelloOboe is more complicated than all of samples listed above. It creates an audio stream onResume() and deletes it onPause(). It also closes the current stream and reopens a new stream whenever PlaybackEngine.setChannelCount, PlaybackEngine.setAudioDeviceId, or PlaybackEngine.setAudioApi is called. This is clearly the largest sample despite the name.

I recommend making changes to LiveEffect for a Java sample. For the wiki, we can list out 4 samples in the following chart:

| Type of sample             | Kotlin      | Java       |
|----------------------------|-------------|------------|
| Create engine at on resume | DrumThumper | Soundboard |
| Dynamically create engine  | MinimalOboe | LiveEffect |

**Comment by philburk** [comment]
I agree LiveEffect would be a good candidate for a background task.

Re modifying SoundBoard
>having pads that do not play audio at the beginning feels like a bug.

Ideally the open should just take 100-200 msec. The user should not notice the unresponsive pads. But if the audioserver dies and takes 10 seconds to recover then the ANR in the UI thread will kill the app. That is not acceptable. So all apps should really use a background task to avoid ANRs.

**Comment by philburk** [comment]
@mmoczkowski - Do you know Kotlin/Compose? If so can you review to see if this demonstrates proper techniques? I am new to both.

**Comment by robertwu1** [comment]
> @robertwu1 wrote:
> 
> > This file also needs to a line about DrumThumper
> 
> I don't understand. Can you clarify?

It currently lists out 5 samples. It doesn't list out DrumThumper

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+import androidx.compose.runtime.*
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.lifecycle.ViewModel
+import com.example.minimaloboe.ui.theme.SamplesTheme
+import kotlinx.coroutines.*
+
+class MainActivity : ComponentActivity() {
+
+    var mExampleViewModel = ExampleViewModel()
```
Comment (mmoczkowski) [inline samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:34]: ViewModels should only be instantiated by `ViewModelProvider` ([see here](https://developer.android.com/topic/libraries/architecture/viewmodel#viewmodel-with-dependencies)).  Easiest way to do so:

1. Declare `androidx.activity:activity-ktx` dependency in Gradle
2. Use delegated property to instantiate VM: `val exampleViewModel: ExampleViewModel by viewModels()`

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                    modifier = Modifier.fillMaxSize(),
+                    color = MaterialTheme.colors.background
+                ) {
+                    MainControls(mExampleViewModel)
+                }
+            }
+        }
+    }
+
+    override fun onStop() {
```
Comment (mmoczkowski) [inline samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:51]: Overriding lifecycle methods is now discouraged in favor of using `LifecycleObserver`. I'd suggest:

1. Make `AudioPlayer` extend the `DefaultLifecycleObserver`
2. Register lifecycle observer in Activity's `onCreate`: `lifecycle.addObserver(player)`

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        return audioPlayer.startAudio()
+    }
+
+    fun stopAudio(): Int {
+        return audioPlayer.stopAudio()
+    }
+}
+
+@Composable
+fun MainControls(viewModel: ExampleViewModel = ExampleViewModel()) {
```
Comment (mmoczkowski) [inline samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:72]: 1. viewModel shouldn't be instantiated like that, but you can do something very similiar with hilt viewmodel integration `viewModel: ExampleViewModel = viewModel()`. See [here](https://developer.android.com/jetpack/compose/libraries#hilt)
2. Composable should provide a modifier argument. See [here](https://chris.banes.dev/always-provide-a-modifier/)

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+    Column {
+        Text(text = "Minimal Oboe!")
+        // START button
+        Button(
+            onClick = {
+                // Start Oboe from a coroutine in case it blocks for too long.
+                // If the AudioServer has died it may take several seconds to recover.
+                // That can cause an ANR if we are starting audio from the main UI thread.
+                GlobalScope.launch {
```
Comment (mmoczkowski) [inline samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt:88]: ViewModel should expose non-suspending functions and launch coroutines itself. `GlobalScope` should be avoided, use `viewModelScope` instead.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 10 above)
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt]: Thanks Milosz. Don refactored the app and it now uses LifecycleObserver.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 5 above)
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt]: I tried on Oriole running TM and did not see slow animation.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 6 above)
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt]: Don fixed this.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/AudioPlayer.kt [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    init {
+        // Load the library containing the native code including the JNI functions.
+        System.loadLibrary("minimaloboe")
+    }
+
+    fun setPlaybackEnabled(isEnabled: Boolean) {
+        // Start (and stop) Oboe from a coroutine in case it blocks for too long.
+        // If the AudioServer has died it may take several seconds to recover.
+        // That can cause an ANR if we are starting audio from the main UI thread.
+        GlobalScope.launch {
```
[the remark this answers is not in the document]
Comment (dturner) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/AudioPlayer.kt:41]: I'm going to answer my own question here and say that using `GlobalScope` is nearly always a bad idea. Despite our `AudioPlayer` being a singleton (`object`) which will live for the lifetime of the application, any jobs executed by `AudioPlayer` should be executed within a scope owned by `AudioPlayer`. For this reason I've just submitted a PR which adds a `CoroutineContext` to this class under which jobs can be executed and cancelled if `AudioPlayer` is destroyed (if, for example, someone changed it from `object` to `class`).

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 11 above)
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt]: I will add a modifier argument.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 5 above)
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt]: Let's tweak this later if still a problem.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt (hunk 11 above)
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt]: I think this obsolete with Don's changes.

File: samples/minimaloboe/src/main/java/com/example/minimaloboe/AudioPlayer.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (philburk) [reply samples/minimaloboe/src/main/java/com/example/minimaloboe/AudioPlayer.kt]: Thanks Don. I merged that change.

**Comment by philburk** [comment]
@donturner - Thanks for the coroutine scope patch. Requesting final review and Approval.

File: samples/minimaloboe/src/main/res/values/colors.xml [hunk 14]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <color name="purple_200">#FFBB86FC</color>
+    <color name="purple_500">#FF6200EE</color>
+    <color name="purple_700">#FF3700B3</color>
```
Comment (robertwu1) [inline samples/minimaloboe/src/main/res/values/colors.xml:5]: nit: Seems like the only color defined here that is used is purple_700. Is this ok?

File: samples/minimaloboe/src/main/res/mipmap-anydpi-v26/ic_launcher.xml [hunk 15]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
+    <background android:drawable="@drawable/ic_launcher_background" />
+    <foreground android:drawable="@drawable/ic_launcher_foreground" />
+</adaptive-icon>
```
Comment (robertwu1) [inline samples/minimaloboe/src/main/res/mipmap-anydpi-v26/ic_launcher.xml:5]: Add newlines to end of files

**Review by robertwu1 — APPROVED** [review]

**Review by mmoczkowski — APPROVED** [review]

File: samples/minimaloboe/src/main/res/values/colors.xml (hunk 14 above)
Comment (philburk) [reply samples/minimaloboe/src/main/res/values/colors.xml:5]: This was default code provided by Android Studio.

File: samples/minimaloboe/src/main/res/mipmap-anydpi-v26/ic_launcher.xml (hunk 15 above)
Comment (philburk) [reply samples/minimaloboe/src/main/res/mipmap-anydpi-v26/ic_launcher.xml:5]: More default files provided by Android Studio. Should be fixed in Android Studio.

**Comment by philburk** [comment]
This was merged manually to avoid merge conflicts.
I followed the instructions given by GitHub, which unfortunately did not involve a squash. So it was a messy merge.

**Pull request merged**